src/dotnet-interactive-vscode-common/src/metadataUtilities.ts

```diff
@@ -0,0 +1,316 @@
+
+    return notebookMetadata;
+}
+
+function ensureProperShapeForDocumentKernelInfo(kernelInfo: contracts.DocumentKernelInfo) {
+    if (!kernelInfo.aliases) {
+        kernelInfo.aliases = [];
+    }
+
+    return kernelInfo;
+}
+
+export function getKernelspecMetadataFromIpynbNotebookDocument(notebook: vscodeLike.NotebookDocument): KernelspecMetadata {
+    // defaulting to empty values so we don't mis-represent the document
+    const kernelspecMetadata: KernelspecMetadata = {
+        display_name: '',
+        language: '',
+        name: ''
+    };
+
+    const custom = notebook.metadata.custom;
+    if (typeof custom === 'object') {
+        const metadata = custom.metadata;
+        if (typeof metadata === 'object') {
+            const kernelspec = metadata.kernelspec;
+            if (typeof kernelspec === 'object') {
+                const display_name = kernelspec.display_name;
+                if (typeof display_name === 'string') {
+                    kernelspecMetadata.display_name = display_name;
+                }
+
+                const language = kernelspec.language;
+                if (typeof language === 'string') {
+                    kernelspecMetadata.language = language;
+                }
+
+                const name = kernelspec.name;
+                if (typeof name === 'string') {
+                    kernelspecMetadata.name = name;
+                }
+            }
+        }
+    }
+
+    return kernelspecMetadata;
+}
+
+export function getKernelInfosFromNotebookDocument(notebookDocument: vscodeLike.NotebookDocument): contracts.KernelInfo[] {
+    const notebookDocumentMetadata = getNotebookDocumentMetadataFromNotebookDocument(notebookDocument);
+    const kernelInfos: contracts.KernelInfo[] = notebookDocumentMetadata.kernelInfo.items.map(item => ({
+        // these are the only important ones
+        localName: item.name,
+        aliases: item.aliases,
+        languageName: item.languageName,
+        // these are unused
+        displayName: 'unused',
+        supportedKernelCommands: [],
+        supportedDirectives: []
+    }));
+    return kernelInfos;
+}
+
+export function getKernelspecMetadataFromNotebookDocumentMetadata(notebookDocumentMetadata: NotebookDocumentMetadata): KernelspecMetadata {
+    // these options are hard-coded because this is exactly what we put on disk with `dotnet interactive jupyter install`
+    switch (notebookDocumentMetadata.kernelInfo.defaultKernelName) {
+        case 'fsharp':
+            return {
+                display_name: '.NET (F#)',
+                language: 'F#',
+                name: '.net-fsharp'
+            };
+        case 'pwsh':
+            return {
+                display_name: '.NET (PowerShell)',
+                language: 'PowerShell',
+                name: '.net-pwsh'
+            };
+        case 'csharp':
+        default:
+            return {
+                display_name: '.NET (C#)',
+                language: 'C#',
```

Comment:
I wonder if this should be changed everywhere to `C# Script`.

Reply:
This is used by the jupyter front end to set up the lexer.  Besides, the language is still "C#", but the dialect is "script" vs "project".